### PR TITLE
EGamma custom nano 

### PR DESCRIPTION
**Pull request by swagata87**
#### PR description:
This PR introduces EGamma custom nano for offline electron and photon objects, as discussed in [1] during XPOG workshop.

[1] https://indico.cern.ch/event/1263032/contributions/5450868/attachments/2672569/4633289/xpog_EGM.pdf

#### PR validation:

Checked with [2] for 12434.0

[2]
```
cmsDriver.py step3 -s RAW2DIGI,L1Reco,RECO,RECOSIM,PAT,NANO,VALIDATION:@standardValidation+@miniAODValidation,DQM:@standardDQM+@ExtraHLT+@miniAODDQM+@nanoAODDQM --conditions auto:phase1_2023_realistic --datatier GEN-SIM-RECO,MINIAODSIM,NANOAODSIM,DQMIO -n 10 --eventcontent RECOSIM,MINIAODSIM,NANOAODSIM,DQM --geometry DB:Extended --era Run3_2023 --customise PhysicsTools/NanoAOD/nano_cff.addExtraEGammaVarsCustomize --filein file:step2.root
```

PS: Parsing of remainder was implemented as it was needed for some variables (https://cms-talk.web.cern.ch/t/issue-with-parsing-modulo-operator-in-nano/28519)

(With thanks to Giovanni Petrucciani and Riccardo Salvatico for help and useful discussions)

### CODE REVIEW

**Comment by swagata87** [comment]
type egamma

File: PhysicsTools/NanoAOD/python/nano_cff.py [hunk 1]
```diff
@@ -283,3 +283,12 @@ def nanoL1TrigObjCustomize(process):
 def nanoL1TrigObjCustomizeFull(process):
     process.nanoTableTaskCommon.add(process.l1TablesTask)
     return process
+
+def addExtraEGammaVarsCustomize(process):
```
Comment (simonepigazzini) [inline PhysicsTools/NanoAOD/python/nano_cff.py:287]: Hi @swagata87,

can you please move the customize function in a separate cff (egamma_custom_cff.py)? This to be in line with the changes that will be introduced by: https://github.com/cms-sw/cmssw/pull/42238

Comment (swagata87) [reply PhysicsTools/NanoAOD/python/nano_cff.py]: Done,
with this change, the customise command in cmsDriver would be `--customise PhysicsTools/NanoAOD/egamma_custom_cff.addExtraEGammaVarsCustomize`

**Comment by simonepigazzini** [comment]
Thank you, can you also dynamically add entries to the NanoDQM activated through the same function? eventually we will add a test workflow in a separate PR

**Comment by swagata87** [comment]
> Thank you, can you also dynamically add entries to the NanoDQM activated through the same function? eventually we will add a test workflow in a separate PR

hello @simonepigazzini it seems it's not possible to add histograms to NanoDQM through the same function. I will try to do it in the same way as done for jme and gen DQM, ie by introducing new files like [nanojmeDQM_cff.py](https://github.com/cms-sw/cmssw/blob/master/PhysicsTools/NanoAOD/python/nanojmeDQM_cff.py).

**Comment by swagata87** [comment]
added histograms for extra variables using a process modifier. 
Checked using the following commands:
```
cmsDriver.py step3  -s RAW2DIGI,L1Reco,RECO,RECOSIM,PAT,NANO,VALIDATION:@standardValidation+@miniAODValidation,DQM:@standardDQM+@ExtraHLT+@miniAODDQM+@nanoAODDQM --conditions auto:phase1_2023_realistic --datatier GEN-SIM-RECO,MINIAODSIM,NANOAODSIM,DQMIO -n 10 --eventcontent RECOSIM,MINIAODSIM,NANOEDMAODSIM,DQM --geometry DB:Extended --era Run3_2023 --filein  file:step2.root --customise PhysicsTools/NanoAOD/egamma_custom_cff.addExtraEGammaVarsCustomize --procModifiers egmCustomNano  --fileout file:step3.root 

cmsDriver.py step4  -s HARVESTING:@standardValidation+@standardDQM+@ExtraHLT+@miniAODValidation+@miniAODDQM+@nanoAODDQM --conditions auto:phase1_2023_realistic --mc  --geometry DB:Extended --scenario pp --filetype DQM --era Run3_2023 -n 100  --filein file:step3_inDQM.root --procModifiers egmCustomNano  --fileout file:step4.root 
```

**Comment by smuzaffar** [comment]
Milestone for this pull request has been moved to CMSSW_14_0_X.Please open a backport if it should also go in to CMSSW_13_3_X.

**Comment by swagata87** [comment]
Is there anything else needed for this to proceed?
I believe I have done what was asked.

**Comment by mandrenguyen** [comment]
+1

**Comment by vlimant** [comment]
configuring the custom nano with an era modifier is not the recommender way.
please adopt, either `--step NANO:<my_cff>` or `--customize <my_customize_function>` or `--step NANO:@EGM`  (#42238)

**Comment by vlimant** [comment]
introducing a test workflow in the nano test matrix would also help with making sure this stays maintained (see #43065)

**Comment by RSalvatico** [comment]
I will make a few comments on this PR to revamp it and hopefully help completing it.

File: PhysicsTools/NanoAOD/python/photons_cff.py [hunk 2]
```diff
@@ -252,7 +255,54 @@ def make_bitmapVID_docstring(id_modules_working_points_pset):
...
+
+photonTable.variables = cms.PSet(
+    _phoVarsBase
+)
+
+photonExtraTable = photonTable.clone()
+
+_phoVarsExtra = cms.PSet(
+    r9Frac = Var("r9()",float,doc="Fractional R9 of the supercluster",precision=10),
+    superclusterEta = Var("superCluster().eta()",float,doc="supercluster eta",precision=10),
```
Comment (RSalvatico) [inline PhysicsTools/NanoAOD/python/photons_cff.py:269]: This was added by https://github.com/cms-sw/cmssw/pull/43198, so it can be removed from here.

**Comment by RSalvatico** [comment]
> configuring the custom nano with an era modifier is not the recommender way. please adopt, either `--step NANO:<my_cff>` or `--customize <my_customize_function>` or `--step NANO:@EGM` (#42238)

This can be done with something like:

https://github.com/RSalvatico/cmssw/blob/a6210dafa1c8d14fbc9e9b8832b5231b7b3dd7c1/PhysicsTools/NanoAOD/python/autoNANO.py#L36-L38

I've tested it by changing `-s NANO` to `-s NANO:@EGM`:

`cmsDriver.py step3  -s RAW2DIGI,L1Reco,RECO,RECOSIM,PAT,NANO:@EGM,VALIDATION:@standardValidation+@miniAODValidation,DQM:@standardDQM+@ExtraHLT+@miniAODDQM+@nanoAODDQM --conditions auto:phase1_2023_realistic --datatier GEN-SIM-RECO,MINIAODSIM,NANOAODSIM,DQMIO -n 10 --eventcontent RECOSIM,MINIAODSIM,NANOAODSIM,DQM --geometry DB:Extended --era Run3_2023 --filein  file:step2.root  --fileout file:step3.root   > step3_TTbar_14TeV+2023.log`

and the new variables were in the output Nano.

File: PhysicsTools/NanoAOD/python/nanoDQM_cff.py [hunk 3]
```diff
@@ -18,6 +18,131 @@
...
+from Configuration.ProcessModifiers.egmCustomNano_cff import egmCustomNano
+(egmCustomNano).toModify(
+    nanoDQM.vplots.Electron,
+    plots = _Electron_extra_plots
+)
+(egmCustomNano).toModify(
+    nanoDQM.vplots.Photon,
+    plots = _Photon_extra_plots
+)
+
```
Comment (RSalvatico) [inline PhysicsTools/NanoAOD/python/nanoDQM_cff.py:145]: These can be removed (and so can `egmCustomNano_cff.py`) and I believe the additional plots can be initialized from within the `addExtraEGammaVarsCustomize` function in `egamma_custom_cff.py` by doing something like:

https://github.com/RSalvatico/cmssw/blob/45caabcfd5167ed973ffb15df03759de176779f3/PhysicsTools/NanoAOD/python/egamma_custom_cff.py#L7-L18

I admit though that I am not sure how to toggle the EGM customization for nanoDQM the same way as NANO:@EGM. Something as trivial as @nanoAODDQM:@EGM does not work, and I don't know if it is even supposed to. Perhaps @vlimant can shed some light on this.

**Comment by vlimant** [comment]
assuming one will always run the DQM step together with the NANO, it is fine to have it in the customise function

**Comment by vlimant** [comment]
looks like the only thing you need to customise, instead of cloning module, duplicate sequence, etc is to create a PSet for the extra variables (one for the table producer and one for the dqm) and extend the existing PSet (I cannot recall the exact syntax to merge two PSet though ; will find out)

**Comment by smuzaffar** [comment]
@swagata87 , over 2.2K commits and 3.8K files changes for this PR do not look right

**Comment by vlimant** [comment]
> I cannot recall the exact syntax to merge two PSet though

`aPSet.setValue(bPset.parameters_())`

**Comment by swagata87** [comment]
> > I cannot recall the exact syntax to merge two PSet though
> 
> `aPSet.setValue(bPSet.parameters_())`

sorry didn't follow it very well; this comment is about which file? just trying to understand where I need to change

File: PhysicsTools/NanoAOD/python/electrons_cff.py [hunk 4]
```diff
@@ -297,7 +297,11 @@ def _get_bitmapVIDForEle_docstring(modules,WorkingPoints):
     src = cms.InputTag("linkedObjects","electrons"),
     name= cms.string("Electron"),
     doc = cms.string("slimmedElectrons after basic selection (" + finalElectrons.cut.value()+")"),
-    variables = cms.PSet(CandVars,
+    variables = cms.PSet(), ## PSet defined below in mode-dependent way
```
Comment (vlimant) [inline PhysicsTools/NanoAOD/python/electrons_cff.py:300]: please leave the content of `variables` the default one, and extend it with the `_eleVarsExtra` and `` ; i.e. modifying `electronTable` on the spot in the customise function

**Comment by vlimant** [comment]
> > > I cannot recall the exact syntax to merge two PSet though
> > 
> > 
> > `aPSet.setValue(bPSet.parameters_())`
> 
> sorry didn't follow it very well; this comment is about which file? just trying to understand where I need to change

I meant this https://github.com/cms-sw/cmssw/pull/42607#discussion_r1489237602

**Comment by swagata87** [comment]
Hi @cms-sw/xpog-l2 , is there anything else needed for this to proceed? We would like to have it soon. So if there is any showstopper, please let us know

File: PhysicsTools/NanoAOD/python/nanoDQM_cff.py [hunk 5]
```diff
@@ -18,6 +18,106 @@
     plots = _boostedTauPlotsV10
 )
 
+## EGamma custom nano
+_Electron_extra_plots = cms.VPSet()
```
Comment (vlimant) [inline PhysicsTools/NanoAOD/python/nanoDQM_cff.py:22]: a more elegant way would be to do  `_Electron_extra_plots = nanoDQM.vplots.Electron.plots.copy()` IMO.
I'd appreciate if you could make this change (independent of the purpose of this PR) every in this file where there is a full copy of a VPSet

**Comment by vlimant** [comment]
this is converging. The remaining item is to have an autoNano entry and a test workflow in the nano matrix : please see #43065 for indications on how to do this

**Comment by vlimant** [comment]
please test

File: PhysicsTools/NanoAOD/python/egamma_custom_cff.py [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+from PhysicsTools.NanoAOD.photons_cff import *
+from PhysicsTools.NanoAOD.photons_cff import _phoVarsExtra
+from PhysicsTools.NanoAOD.NanoAODEDMEventContent_cff import *
+from PhysicsTools.NanoAOD.nanoDQM_cfi import nanoDQM
+from PhysicsTools.NanoAOD.nanoDQM_cff import _Photon_extra_plots, _Electron_extra_plots
+
+def addExtraEGammaVarsCustomize(process):
+    #photon
+    process.photonTable.variables.setValue(_phoVarsExtra.parameters_())
+    process.nanoDQM.vplots.Photon.plots = _Photon_extra_plots
```
Comment (vlimant) [inline PhysicsTools/NanoAOD/python/egamma_custom_cff.py:14]: you might want to add an `if` on the presence of `process.nanoDQM`

File: Configuration/PyReleaseValidation/python/relval_nano.py [hunk 7]
```diff
@@ -175,6 +181,9 @@ def subnext(self):
                              '--conditions':'auto:phase1_2022_realistic'},
                             _NANO_mc])
 
+steps['EGMNano_mc13.2'] = merge([{'-s':'NANO:@EGM', '-n' : '1000'},
```
Comment (vlimant) [inline Configuration/PyReleaseValidation/python/relval_nano.py:184]: you may add `,DQM:@nanoAODDQM` in the steps

**Comment by swagata87** [comment]
upon running `runTheMatrix.py -l 2500.402` , which is `EGMNANOmc132X`,
I'm getting this problem: 
`Looking for productInstanceName: cutBasedElectronID-RunIIIWinter22-V1-veto`. 
@RSalvatico this ID should be present right?

**Comment by RSalvatico** [comment]
> upon running `runTheMatrix.py -l 2500.402` , which is `EGMNANOmc132X`, I'm getting this problem: `Looking for productInstanceName: cutBasedElectronID-RunIIIWinter22-V1-veto`. @RSalvatico this ID should be present right?

Yes. The problem is not that ID in particular - if that commented out, it cannot find the other working points either. The problem is this missing collection in the input datasets:

https://github.com/cms-sw/cmssw/blob/1e2c4980d73549aad8a2c4db1034bd15d2468e0b/PhysicsTools/NanoAOD/python/electrons_cff.py#L72

**Comment by RSalvatico** [comment]
Mmh, I suspect that we are not running the postReco tool and not producing VIDs...

**Comment by RSalvatico** [comment]
Ok, the problem is that we aren't running the sequences corresponding to @PHYS in nano, hence the VIDs are not added to the right electron collection. I see two possible solutions, and I let @vlimant comment on the preferred one.

1. Run `NANO:@PHYS+@EGM` instead of just `NANO:@EGM`. This needs to be changed in the relval_nano but also has to be done every time we need to run `NANO:@EGM` in general.
2. Do something like [*] and then call `@EGMPOG`.

[*]
https://github.com/RSalvatico/cmssw/blob/2496ecd5fb051cf44ee4971d2c1ffe7d6f8d3c1a/PhysicsTools/NanoAOD/python/autoNANO.py#L36-L39

I've tried both options with 
`runTheMatrix.py -l 2500.314 --sites ''`
`runTheMatrix.py -l 2500.315`
`runTheMatrix.py -l 2500.401`

and I've verified that we get the extra branches in NANO if we do something like

```
cmsDriver.py step3  -s RAW2DIGI,L1Reco,RECO,RECOSIM,PAT,NANO:@EGMPOG,VALIDATION:@standardValidation+@miniAODValidation,DQM:@standardDQM+@ExtraHLT+@miniAODDQM+@nanoAODDQM --conditions auto:phase1_2023_realistic --datatier GEN-SIM-RECO,MINIAODSIM,NANOAODSIM,DQMIO -n 10 --eventcontent RECOSIM,MINIAODSIM,NANOAODSIM,DQM --geometry DB:Extended --era Run3_2023 --filein  file:step2.root  --fileout file:step3.root   > step3_TTbar_14TeV+2023.log
```

**Comment by swagata87** [comment]
thanks @RSalvatico. 
@cms-sw/xpog-l2 kindly let us know your preference about this, and we will implement accordingly

**Comment by vlimant** [comment]
you might be able to map @EGM to 

`
{'sequence': '@PHYS',
              'customize' : '@PHYS+PhysicsTools/NanoAOD/egamma_custom_cff.addExtraEGammaVarsCustomize'}
`

**Comment by RSalvatico** [comment]
> you might be able to map @EGM to
> 
> `{'sequence': '@PHYS', 'customize' : '@PHYS+PhysicsTools/NanoAOD/egamma_custom_cff.addExtraEGammaVarsCustomize'}`

Yes, this seems to be working too. Thanks

**Comment by vlimant** [comment]
please test

**Comment by RSalvatico** [comment]
Since the proton prompt 2024 and the 2022-2023 re-recos are going to be done in 140X, we will need a backport. There is also a chance to test the custom nano production this week if we manage to integrate it, I've learned from PPD.

@cms-sw/ppd-l2

**Comment by swagata87** [comment]
if there is no other comment in this PR, and if people consider it ready to be signed then I can squash all commits and also make a backport

**Comment by vlimant** [comment]
enable nano

**Comment by vlimant** [comment]
please test

**Comment by vlimant** [comment]
+1

**Comment by swagata87** [comment]
Squashed the commits and opened backport PR (https://github.com/cms-sw/cmssw/pull/44255)

File: Configuration/PyReleaseValidation/python/relval_nano.py [hunk 8]
```diff
@@ -240,12 +249,15 @@ def subnext(self):
 workflows[_wfn()] = ['muDPGNANO130Xrun3', ['ZMuSkim2023DRAWRECO13.0', 'muDPGNANO_data13.0']]
 workflows[_wfn()] = ['muDPGNANOBkg130Xrun3', ['ZeroBias2023DRAW13.0', 'muDPGNANOBkg_data13.0']]
 workflows[_wfn()] = ['muPOGNANO_data13.0', ['MuonEG2023MINIAOD13.0', 'muPOGNANO_data13.0']]
+workflows[_wfn()] = ['EGMNANOdata130Xrun3', ['MuonEG2023MINIAOD13.0', 'EGMNano_data13.0']]
+workflows[_wfn()] = ['EGMNANOmc130X', ['TTBarMINIAOD13.0', 'EGMNano_mc13.0']]
```
Comment (vlimant) [inline Configuration/PyReleaseValidation/python/relval_nano.py:253]: please move this up to below `NANOmc130X` above

**Comment by vlimant** [comment]
thanks

**Comment by vlimant** [comment]
please test

**Comment by vlimant** [comment]
+1

**Comment by mandrenguyen** [comment]
+reconstruction
Relates to nano only, so if xpog is happy, we're happy.

**Comment by subirsarkar** [comment]
+Upgrade

**Comment by AdrianoDee** [comment]
+pdmv
(wf-wise nothing to comment)

**Comment by rappoccio** [comment]
+1

**Pull request merged**